File: 0692-top-k-frequent-words/0692-top-k-frequent-words.cpp
class Solution {
public:
    
    class compare {
        public:
            bool operator()(const pair<string, int> &p1, const pair<string, int> &p2) {
                if(p1.second == p2.second)
                    return p1.first < p2.first;
                return p1.second > p2.second;
            }
    };
    
    vector<string> topKFrequent(vector<string>& words, int k) {
        unordered_map<string, int> m;
        for(string word : words) m[word]++;

        priority_queue<pair<string, int>, vector<pair<string, int>>, compare> pq;
        for(auto item : m) {
            if(pq.size() == k) {
                if(pq.top().second < item.second || (pq.top().second == item.second && pq.top().first > item.first)) {
                    pq.pop();
                    pq.push({item.first, item.second});
                }
            } else
                pq.push({item.first, item.second});
        }
        
        vector<string> ans;
        while(!pq.empty()) {
            ans.push_back(pq.top().first);
            pq.pop();
        }
        
        reverse(ans.begin(), ans.end());
        return ans;
    }
};

File: 0692-top-k-frequent-words/0692-top-k-frequent-words.java
class Solution {
    
    class Pair implements Comparable<Pair> {
        
        String word;
        int count;
        
        Pair(String word, int count) {
            this.word = word;
            this.count = count;
        }
        
        public int compareTo(Pair x) {
            
            if(count == x.count)
                return x.word.compareTo(word);
            
            return count - x.count;
        }
        
    }
    
    public List<String> topKFrequent(String[] words, int k) {
        
        HashMap<String, Integer> map = new HashMap<>();
        PriorityQueue<Pair> heap = new PriorityQueue<>();
        
        List<String> ans = new ArrayList<>();
        
        for(String word : words)
            map.put(word, map.getOrDefault(word, 0) + 1);
        
        for(String word : map.keySet()) {
            if(heap.size() == k) {
                if(heap.peek().count < map.get(word) || (heap.peek().count == map.get(word) && (word.compareTo(heap.peek().word)) < 0)) {
                    heap.remove();
                    heap.add(new Pair(word, map.get(word)));
                }
            }
            else
                heap.add(new Pair(word, map.get(word)));
        }

        
        while(k --> 0) {
            Pair x = heap.remove();
            ans.add(x.word);
        }
        Collections.reverse(ans);
        
        return ans;
    }
}